### PR TITLE
DeiT (small) on ADNI Dataset | s4744343

### PR DESCRIPTION
**Pull request by mashamegul**
This pull request introduces the foundational codebase for classifying Alzheimer's disease using the ADNI dataset and the Data-efficient Image Transformer (DeiT) model. The accuracy on test data of this model has been developed to be over 80%.

### Key Files
- **train.py**: The training script that includes model setup, data loading, and training loop. Implements gradient accumulation, learning rate scheduling, early stopping, and automatic mixed precision for efficient training.
- **predict.py**: Script for making predictions on individual MRI scans, outputting a classification (NC or AD) with confidence scores. This file also supports batch predictions over test data for bulk inference.
- **modules.py**: Contains the DeiT model implementation along with essential helper classes, including multi-layer perceptron (MLP), global filter layers, and the forward pass structure. This modular setup supports easy customisation and potential integration of additional transformer layers.
- **dataset.py**: Defines the ADNIDataset class responsible for loading and preprocessing the ADNI dataset, along with data transformations specific to training and testing. Ensures MRI data is correctly structured for input into the DeiT model.
- **images/**: Folder containing sample images of the MRI scans, processed input data, and visualisations of model performance, such as confusion matrices and loss curves.
- **README.md**: Comprehensive documentation detailing the project objectives, dataset structure, preprocessing methods, model architecture choices, experimental comparisons with other transformer models, and a step-by-step guide for running the code. Includes insights into the efficacy of different transformers in achieving robust Alzheimer's classification.

**Model Architecture:**
- DeiT Small model setup using pretrained weights from ImageNet.
- Configured for binary classification to distinguish between Normal Control (NC) and Alzheimer's Disease (AD) MRI scans.

### CODE REVIEW

**Comment by shaivikaaaa** [comment]
@gayanku

**Comment by gayanku** [comment]
**This is an initial inspection, no action is required at this point**
Uses a pretrained imageNet model for ADNI classification.

**Comment by aniketgupta17** [comment]
Observational Feedback 

File Organizing: Well-organized files.

Pull Request :
Correctly generated to Topic Recognition Branch .
The Description talks about the File Structure Clearly , Add about achieved Results .  

Problem Solving:

The algorithm solves the problem appropriately   . 
Accuracy in testing dataset:0.8142
Model and functions:

It correctly uses PyTorch to construct the functions.
Code design: Good.

Code comment and docstring:

Sufficient Code Commenting . 
Missing function docstrings
Difficulty: Hard.

Additional Comments:
Meaningful commits.
Good ReadMe design. However, It would be better if a conclusion and future improvements  appeared in the report.

**Comment by mashamegul** [comment]
# Alzheimer's Classification with ViT, using DeiT small

This pull request introduces the foundational codebase and report for classifying Alzheimer's disease using the ADNI dataset and the Data-efficient Image Transformer (DeiT) model, with a target accuracy of above 80% on the test set in Project 5 for COMP3710 A3:
```
Classify Alzheimer’s disease (normal and AD) of the ADNI brain data using one of the latest
vision transformers such as the GFNet set having a minimum accuracy of 0.8 on the test set. 
[Hard Difficulty]
```

### Key Achieved Results
- **Test Accuracy**: Achieved 81.42% accuracy on the test set, surpassing the target for effective differentiation between Normal Control (NC) and Alzheimer's Disease (AD) cases.
- **Performance Metrics**: Confusion matrices and loss curves are included to illustrate the model’s performance, showing clear differentiation between NC and AD cases.
- **Expanded Analysis and Evaluation**: Added in-depth sections on analysis, evaluation, model limitations, and potential extensions for comprehensive insights into the model’s performance and areas for improvement.

### Key Files
- **`train.py`**: The training script, including model setup, data loading, and the training loop. Implements gradient accumulation, learning rate scheduling, early stopping, and automatic mixed precision.
- **`predict.py`**: A script for predicting classifications (NC or AD) with confidence scores for individual or batch MRI scans.
- **`modules.py`**: Contains the DeiT model implementation and helper classes, supporting customisation and integration of additional layers.
- **`dataset.py`**: Defines the ADNIDataset class for loading and preprocessing the ADNI dataset, ensuring compatibility with the DeiT model.
- **`images/`**: Contains sample MRI images and performance visualisations, such as confusion matrices and loss curves.
- **`README.md`**: Comprehensive documentation, including project objectives, dataset structure, preprocessing, model architecture, and a step-by-step guide for running the code. Expanded to include analysis of the model's performance and possible improvements.

### Additional Improvements Based on Feedback
- **Detailed Analysis and Evaluation Sections**: Added sections on in-depth analysis, evaluation of results, limitations, and suggested improvements/extensions for the model.
- **Enhanced Code Comments and Docstrings**: Updated all Python files with detailed comments and docstrings for clarity and documentation of functions and classes.

### Model Architecture
- **DeiT Small**: Configured with pretrained weights from ImageNet for binary classification between NC and AD MRI scans. The modular design supports experimentation with larger DeiT models for potential accuracy gains.

**Comment by mashamegul** [comment]
Hi, the above commit replaces an earlier R&D plot with the correct early stopping plot used in the report. This ensures alignment with the discussed results and conclusions

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting missing. Missing doc string.</td><td>-1</td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. ViT (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by gayanku** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by mashamegul** [comment]
Hi @gayanku,

Thank you for the feedback. I was a bit confused and had assumed that the 2 marks associated with implementing feedback would be given based on @aniketgupta17 's comments, which is why I implemented those changes. Discussions on Ed suggested that lost marks could be recovered through this, alongside the additional 2 marks.

Additionally, I have now updated the branch for the PR correctly. Sorry for the earlier oversight.

Thanks,
Masham
